validator_client/validator_services/src/sync.rs

```diff
@@ -521,129 +522,316 @@ pub async fn fill_in_aggregation_proofs<T: SlotClock + 'static, E: EthSpec>(
-                continue;
-            }
+        // For distributed mode
+        if duties_service.distributed {
+            let mut partial_proofs = Vec::new();
 
-            let subnet_ids = match duty.subnet_ids::<E>() {
-                Ok(subnet_ids) => subnet_ids,
-                Err(e) => {
-                    crit!(
-                        log,
-                        "Arithmetic error computing subnet IDs";
-                        "error" => ?e,
-                    );
-                    continue;
-                }
-            };
+            for (_validator_start_slot, duty) in pre_compute_duties {
+                // Proofs are already known at this slot for this validator.
+
+                let subnet_ids = match duty.subnet_ids::<E>() {
+                    Ok(subnet_ids) => subnet_ids,
+                    Err(e) => {
+                        crit!(
+                            log,
+                            "Arithmetic error computing subnet IDs";
+                            "error" => ?e,
+                        );
+                        continue;
+                    }
+                };
 
-            // Create futures to produce proofs.
-            let duties_service_ref = &duties_service;
-            let futures = subnet_ids.iter().map(|subnet_id| async move {
                 // Construct proof for prior slot.
                 let proof_slot = slot - 1;
 
-                let proof = match duties_service_ref
-                    .validator_store
-                    .produce_sync_selection_proof(&duty.pubkey, proof_slot, *subnet_id)
-                    .await
-                {
-                    Ok(proof) => proof,
-                    Err(ValidatorStoreError::UnknownPubkey(pubkey)) => {
-                        // A pubkey can be missing when a validator was recently
-                        // removed via the API.
+                // Create futures for all subnet IDs for this validator
+                for subnet_id in subnet_ids {
+                    let duties_service = duties_service.clone();
+                    let duty = duty.clone();
+                    let subnet_id = *subnet_id;
+
+                    // Store all partial sync selection proofs for this slot in partial_proofs so that it can be sent together
+                    partial_proofs.push(async move {
+                        // Produce partial selection proof
+                        let sync_selection_proof = duties_service
+                            .validator_store
+                            .produce_sync_selection_proof(
+                                &duty.pubkey,
+                                proof_slot,
+                                subnet_id.into(),
+                            )
+                            .await;
+
+                        match sync_selection_proof {
+                            Ok(proof) => {
+                                let sync_committee_selection = SyncCommitteeSelection {
```

Comment:
Could also log the partial proof here (to make sure it's different from the full proof)